src/Development/Rattle/Profile.hs

```diff
@@ -0,0 +1,252 @@
+{-# LANGUAGE ScopedTypeVariables, RecordWildCards, TupleSections, ViewPatterns #-}
+
+module Development.Rattle.Profile(
+  constructGraph, Graph(..), dotStringOfGraph,
+  graphData, writeProfile
+  ) where
+
+import Development.Rattle.Server
+import Development.Rattle.Types
+import Development.Rattle.Hash
+import Development.Rattle.Shared
+import Control.Monad
+import System.IO.Unsafe(unsafeInterleaveIO)
+import Data.Maybe as Maybe
+import Data.List as List
+import qualified Data.ByteString.Lazy.Char8 as LBS
+import General.Template
+import General.Paths
+import Data.HashMap.Strict as Map
+import Data.HashSet as Set
+import System.Time.Extra
+import Numeric.Extra
+
+-- edge is directed based on order cmd were listed in script
+-- end1 was listed before end2. helps determine read/write hazards
+data Edge = Edge {end1 :: (Cmd, [Trace Hash])
+                 ,end2 :: (Cmd, [Trace Hash])
+                 ,hazard :: Maybe Hazard
+                 }
+
+data Graph = Graph {nodes :: [(Cmd, [Trace Hash])]
+                   ,edges :: [Edge]
+                   }
+
+instance Show Edge where
+  show (Edge e1 e2 Nothing) = showCmd (fst e1) ++ " -> " ++ showCmd (fst e2)
+  show (Edge e1 e2 (Just h)) = showCmd (fst e1) ++ " -> " ++ showCmd (fst e2) ++ " [ label=\"" ++ show h ++ "\" ];"
+
+getCmdsTraces :: RattleOptions -> IO [(Cmd,[Trace Hash])]
+getCmdsTraces options@RattleOptions{..} = withShared rattleFiles$ \shared -> do
+  cmds <- maybe (return []) (getSpeculate shared) rattleSpeculate
+  fmap (takeWhile (not . List.null . snd)) $ forM cmds $ \x -> (x,) <$> unsafeInterleaveIO (getCmdTraces shared x)
```

Comment:
Why the unsafeInterleaveIO?

Reply:
I copied this bit of code from withRattle; so, if that is unnecessary I can remove it.

Reply:
For `withRattle` it means that it doesn't bother loading anything where it isn't later used. I suggest skipping that here since your analysis needs everything (I think).